source/common/http/header_map_impl.h

```diff
@@ -119,14 +119,85 @@ class HeaderMapImpl : public HeaderMap {
     ALL_INLINE_HEADERS(DEFINE_INLINE_HEADER_STRUCT)
   };
 
+  /**
+   * List of HeaderEntryImpl that keeps the pseudo headers (key starting with ':') in the front
+   * of the list (as required by nghttp2) and otherwise maintains insertion order.
+   */
+  class HeaderList {
+  public:
+    HeaderList() : pseudo_headers_end_(headers_.end()) {}
+
+    std::list<HeaderEntryImpl>::iterator insert(HeaderString&& key, HeaderString&& value) {
```

Comment:
These 3 insert functions are basically the same. There must be some way to combine them using templates or otherwise?

Reply:
Yes, will push an update that does this.